### DISCUSSION TITLE
Auto-instrumenting application code?

### DISCUSSION THREAD

**Discussion by adamk9k** [open-telemetry/opentelemetry-ruby-contrib#746] [Q&A] [answered]
Hey quick question, 

reading the available instrumenters and how they're generating spans, it seems to me there's no way to automatically instrument application code, am I correct in that? 

For instance, in this patch for active record 

https://github.com/open-telemetry/opentelemetry-ruby-contrib/blob/main/instrumentation/active_record/lib/opentelemetry/instrumentation/active_record/patches/persistence.rb

I see that every AR method is patched to create a child span. Am I understanding correctly that in order to instrument custom application code, I'd need to pepper the codebase with similar child-span creations, there isn't something that does it for me automatically, right?

For instance, if I have 
```ruby
class SomeController < ApplicationController
  def index
    data = DataCalculator.run!

    render json: data
  end
end
```

and in turn the `DataCalculator.run!` calls 3 other services, I'd have to manually instrument all 3 dependencies as well as the controller action and the `DataCalculator#run` method as well. 

Thanks in advance, just want to make sure I'm not missing something ;)

**Comment by arielvalentin** [comment, marked as answer]
You are correct. 

We do not have any instrumentation that generates spans for _custom application code_. 

That you must instrument yourself.

**Reply by adamk9k** [reply]
Thank you for the quick answer 👍
